src/client/serialize/daos_serialize.c

```diff
@@ -1175,3 +1245,1833 @@ daos_cont_deserialize_attrs(char *filename, uint64_t *num_attrs,
+			rc = -DER_MISC;
+			D_ERROR("Failed to get size of type in single record datatype\n");
+			D_GOTO(out, rc);
+		}
+		D_ALLOC(single_data, single_tsize);
+		if (single_data == NULL)
+			D_GOTO(out, rc = -DER_NOMEM);
+		memcpy(single_data, rec_single_val->p, rec_single_val->len);
+
+		/* set iod values */
+		iod.iod_type  = DAOS_IOD_SINGLE;
+		iod.iod_size  = single_tsize;
+		iod.iod_nr    = 1;
+		iod.iod_recxs = NULL;
+		iod.iod_name  = aiov;
+
+		/* set sgl values */
+		sgl.sg_nr     = 1;
+		sgl.sg_nr_out = 0;
+		sgl.sg_iovs   = &iov;
+		d_iov_set(&iov, single_data, single_tsize);
+
+		/* update fetched recx values and place in destination object */
+		rc = daos_obj_update(*oh, DAOS_TX_NONE, 0, &diov, 1, &iod, &sgl, NULL);
+		if (rc != 0) {
+			D_ERROR("Failed to update object: "DF_RC, DP_RC(rc));
+			D_GOTO(out, rc);
+		}
+		(*bytes_written) += single_tsize;
+		D_FREE(single_data);
+	}
+	(*total_akeys)++;
+out:
+	if (rx_dset >= 0)
+		H5Dclose(rx_dset);
+	if (rx_dspace >= 0)
+		H5Sclose(rx_dspace);
+	if (rx_dtype >= 0)
+		H5Tclose(rx_dtype);
+	if (plist >= 0)
+		H5Pclose(plist);
+	D_FREE(dset_name);
+	D_FREE(single_data);
+	return rc;
+}
+
+static int
+cont_deserialize_keys(struct hdf5_args *hdf5, uint64_t *total_dkeys_this_oid, uint64_t *dk_off,
+		      daos_handle_t *oh, hsize_t dkey_dims[], hsize_t akey_dims[],
+		      int *total_dkeys, int *total_akeys, uint64_t *bytes_written)
+{
+	int		rc = 0;
+	int		j = 0;
+	daos_key_t	diov;
+	char		dkey[ENUM_KEY_BUF] = {0};
+	uint64_t	ak_off = 0;
+	uint64_t	ak_next = 0;
+	uint64_t	total_akeys_this_dkey = 0;
+	int		k = 0;
+	hvl_t		*dkey_val;
+	hvl_t		*rec_kv_val;
+	daos_size_t	kv_single_size = 0;
+
+	for (j = 0; j < *total_dkeys_this_oid; j++) {
+		memset(&diov, 0, sizeof(diov));
+		memset(dkey, 0, sizeof(dkey));
+		dkey_val = &(hdf5->dkey_data)[*dk_off + j].dkey_val;
+		rec_kv_val = &(hdf5->dkey_data)[*dk_off + j].rec_kv_val;
+		memcpy(dkey, dkey_val->p, dkey_val->len);
+		d_iov_set(&diov, (void *)dkey_val->p, dkey_val->len);
+		ak_off = hdf5->dkey_data[*dk_off + j].akey_offset;
+		ak_next = 0;
+		total_akeys_this_dkey = 0;
+		if (*dk_off + j + 1 < (int)dkey_dims[0]) {
+			ak_next = hdf5->dkey_data[(*dk_off + j) + 1].akey_offset;
+			total_akeys_this_dkey = ak_next - ak_off;
+		} else if (*dk_off + j == ((int)dkey_dims[0] - 1)) {
+			total_akeys_this_dkey = ((int)akey_dims[0]) - ak_off;
+		}
+
+		/* if rec_kv_val.len != 0 then skip akey iteration, we can write data back
+		 * into DAOS using the daos_kv.h API using just oid, dkey, and key value
+		 * (stored in dkey dataset)
+		 */
+
+		/* run daos_kv_put on rec_kv_val (dkey val) and key (dkey) */
+		/* skip akey iteration for DAOS_OF_KV_FLAT objects
+		 */
+		if (rec_kv_val->len > 0) {
+			kv_single_size = rec_kv_val->len;
+			rc = daos_kv_put(*oh, DAOS_TX_NONE, 0, dkey, kv_single_size,
+					 rec_kv_val->p, NULL);
+			if (rc != 0) {
+				D_ERROR("failed to write kv object "DF_RC"\n", DP_RC(rc));
+				D_GOTO(out, rc);
+  			}
```

Comment:
(style)  code indent should use tabs where possible
(style)  please, no space before tabs